File: src/main/java/org/cytoscape/biogwplugin/internal/cache/BGCache.java
package org.cytoscape.biogwplugin.internal.cache;

import org.cytoscape.biogwplugin.internal.BGServiceManager;
import org.cytoscape.biogwplugin.internal.query.BGFetchRelationTypesQuery;
import org.cytoscape.biogwplugin.internal.query.BGNode;
import org.cytoscape.biogwplugin.internal.query.BGQueryBuilderModel;
import org.cytoscape.biogwplugin.internal.query.QueryTemplate;
import org.cytoscape.model.CyNetwork;
import org.cytoscape.model.CyNode;
import org.cytoscape.model.CyRow;
import org.cytoscape.model.CyTable;
import org.cytoscape.work.TaskIterator;

import java.io.IOException;
import java.io.InputStream;
import java.net.URL;
import java.net.URLConnection;
import java.util.*;

/**
 * Created by sholmas on 24/03/2017.
 */
public class BGCache {

    private HashMap<String, BGNode> nodeCache = new HashMap<>();
    private HashMap<String, String> relationTypes = new HashMap<>();
    private BGServiceManager serviceManager;
    private HashMap<String, QueryTemplate> queryTemplateHashMap;

    public BGCache(BGServiceManager serviceManager) {
        this.serviceManager = serviceManager;

        // Load initial data

        // Load the XML file

        // Load a list of Relation Types
        BGFetchRelationTypesQuery relationTypesQuery = new BGFetchRelationTypesQuery(serviceManager.SERVER_PATH);
        Runnable callback = () -> {
            this.relationTypes = relationTypesQuery.returnData;
        };
        relationTypesQuery.addCallback(callback);
        relationTypesQuery.run();

        loadXMLFileFromServer();

    }

    private void loadXMLFileFromServer() {
        try {
            URL queryFileUrl = new URL("https://dl.dropboxusercontent.com/u/32368359/BiogatewayQueries.xml");
            URLConnection connection = queryFileUrl.openConnection();
            InputStream is = connection.getInputStream();
            queryTemplateHashMap = BGQueryBuilderModel.parseXMLFile(is);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public HashMap<String, QueryTemplate> getQueryTemplateHashMap() {
        if (queryTemplateHashMap != null) {
            return queryTemplateHashMap;
        } else {
            loadXMLFileFromServer();
            return queryTemplateHashMap;
        }
    }

    public BGNode getNodeWithURI(String uri) {

        BGNode node = nodeCache.get(uri);

        if (node == null) {
            System.out.println("BGCache Warning: Node "+uri+" not found. Attempting to fetch it from the current CyNetworks.");
            for (CyNetwork network : serviceManager.getNetworkManager().getNetworkSet()) {
                node = fetchNodeFromCyNetwork(uri, network);
                if (node != null) {
                    nodeCache.put(uri, node);
                    return node;
                }
            }
        }

        return node;
    }

    public ArrayList<BGNode> getNodesWithURIs(Collection<String> uris) {
        ArrayList<BGNode> nodes = new ArrayList();
        for (String uri : uris) {
            BGNode node = getNodeWithURI(uri);
            if (node != null) {
                nodes.add(node);
            }
        }
        return nodes;
    }

    public void addNode(BGNode node) {
        nodeCache.put(node.URI, node);
    }

    public void debug_printCache() {
        for (String key : nodeCache.keySet()) {
            System.out.println(nodeCache.get(key).URI);
        }
    }

    private BGNode fetchNodeFromCyNetwork(String nodeUri, CyNetwork network) {
        CyTable table = network.getDefaultNodeTable();
        Set<CyNode> nodes = getNodesWithValue(network, table, "identifier uri", nodeUri);
        if (nodes.size() == 1) {
            // TODO: Make sure that the identifier uri is found!
            CyNode cyNode = nodes.iterator().next();
            String commonName = table.getRow(cyNode.getSUID()).get("name", String.class);
            BGNode bgNode = new BGNode(nodeUri);
            bgNode.commonName = commonName;
            return bgNode;
        } else if (nodes.size() > 1) {
            System.out.println("BGCache Warning: CyNetwork inconsistency: Found multiple nodes with same URI in the same network!");
            return null;
        } else {
            System.out.println("BGCache Warning: Did not find missing node in CyNetwork table either.");
            return null;
        }
    }

    private static Set<CyNode> getNodesWithValue(CyNetwork network, CyTable table, String columnName, Object value) {
        final Collection<CyRow> matchingRows = table.getMatchingRows(columnName, value);
        final Set<CyNode> nodes = new HashSet<CyNode>();
        final String primaryKeyColname = table.getPrimaryKey().getName();
        for (final CyRow row : matchingRows)
        {
            final Long nodeId = row.get(primaryKeyColname, Long.class);
            if (nodeId == null)
                continue;
            final CyNode node = network.getNode(nodeId);
            if (node == null)
                continue;
            nodes.add(node);
        }
        return nodes;
    }

    public HashMap<String, String> getRelationTypes() {
        return relationTypes;
    }
    public String getNameForRelationType(String relationTypeURI) {
        return relationTypes.get(relationTypeURI);
    }
}

File: src/main/java/org/cytoscape/biogwplugin/internal/old/cache/BGCache.kt
package org.cytoscape.biogwplugin.internal.old.cache

import org.cytoscape.biogwplugin.internal.BGServiceManager
import org.cytoscape.biogwplugin.internal.old.query.BGNode
import org.cytoscape.biogwplugin.internal.old.query.BGQueryBuilderModel
import org.cytoscape.biogwplugin.internal.old.query.QueryTemplate
import org.cytoscape.model.CyNetwork
import org.cytoscape.model.CyNode
import org.cytoscape.model.CyTable

import java.io.IOException
import java.net.URL
import java.util.*
import kotlin.collections.HashMap

/**
 * Created by sholmas on 24/03/2017.
 */
class BGCache(private val serviceManager: BGServiceManager) {

    private val nodeCache = HashMap<String, BGNode>()
    var relationTypes = HashMap<String, String>()
        private set
    private var queryTemplateHashMap: HashMap<String, QueryTemplate> = HashMap()

    init {

        // Load initial data

        // Load the XML file

        // Load a list of Relation Types
//        val relationTypesQuery = BGFetchRelationTypesQuery(BGServiceManager.SERVER_PATH)
//        relationTypesQuery.addCallback { this.relationTypes = relationTypesQuery.returnData  }
//        relationTypesQuery.run()
        loadXMLFileFromServer()
    }

    private fun loadXMLFileFromServer() {
        try {
            val queryFileUrl = URL("https://dl.dropboxusercontent.com/u/32368359/BiogatewayQueries.xml")
            val connection = queryFileUrl.openConnection()
            val `is` = connection.getInputStream()
            queryTemplateHashMap = BGQueryBuilderModel.parseXMLFile(`is`)
        } catch (e: IOException) {
            e.printStackTrace()
        }

    }

    fun getQueryTemplateHashMap(): HashMap<String, QueryTemplate> {
        // TODO: Make this asynchronous.
        if (queryTemplateHashMap.size == 0) {
            loadXMLFileFromServer()
        }
        return queryTemplateHashMap
    }

    fun getNodeWithURI(uri: String): BGNode? {
        // TODO: Make this asynchronous.
        var node: BGNode? = nodeCache[uri]
        if (node == null) {
            println("BGCache Warning: Node $uri not found. Attempting to fetch it from the current CyNetworks.")
            for (network in serviceManager.networkManager.networkSet) {
                node = fetchNodeFromCyNetwork(uri, network)
                if (node != null) {
                    nodeCache.put(uri, node)
                    return node
                }
            }
        }
        return node
    }

    fun getNodesWithURIs(uris: Collection<String>): ArrayList<BGNode> {
        // TODO: Make this asynchronous.
        val nodes = ArrayList<BGNode>()
        for (uri in uris) {
            val node = getNodeWithURI(uri)
            if (node != null) {
                nodes.add(node)
            }
        }
        return nodes
    }

    fun addNode(node: BGNode) {
        nodeCache.put(node.URI, node)
    }


    private fun fetchNodeFromCyNetwork(nodeUri: String, network: CyNetwork): BGNode? {
        val table = network.defaultNodeTable
        val nodes = getNodesWithValue(network, table, "identifier uri", nodeUri)
        if (nodes.size == 1) {
            // TODO: Make sure that the identifier uri is found!
            val cyNode = nodes.iterator().next()
            val commonName = table.getRow(cyNode.suid).get("name", String::class.java)
            val bgNode = BGNode(nodeUri)
            bgNode.commonName = commonName
            return bgNode
        } else if (nodes.size > 1) {
            println("BGCache Warning: CyNetwork inconsistency: Found multiple nodes with same OPTIONAL_URI in the same network!")
            return null
        } else {
            println("BGCache Warning: Did not find missing node in CyNetwork table either.")
            return null
        }
    }

    private fun getNodesWithValue(network: CyNetwork, table: CyTable, columnName: String, value: Any): Set<CyNode> {
        val matchingRows = table.getMatchingRows(columnName, value)
        val nodes = HashSet<CyNode>()
        val primaryKeyColname = table.primaryKey.name
        for (row in matchingRows) {
            val nodeId = row.get(primaryKeyColname, Long::class.java) ?: continue
            val node = network.getNode(nodeId) ?: continue
            nodes.add(node)
        }
        return nodes
    }

    fun getNameForRelationType(relationTypeURI: String): String? {
        return relationTypes[relationTypeURI]
    }
}
